### PR TITLE
IAP: make it possible to optionally provide sandbox url for testing

### LINKED ISSUE #7927: [IAP] Make it possible to optionally provide sandbox URL for testing

### PR DESCRIPTION
**Pull request by kidinov**
Closes: #7927

Do not merge till https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2542 is merged and fluxc hash's changed. You can test it though via local build

### Description

* Adds the possibility to use sandbox base URL for mobile pay APIs
* Initial idea to use fluxc inside the IAP module doesn't work well, because it'll require making it possible to work with local builds as well, that's why providing the network communication now on the app

### Testing instructions

* Build debug build. Make a purchase and notice that network call is not done, but stubbed backed is used
* Build release build. Make a purchase and notice that the network call is not done, and real backed was hit
* add `iap_testing_sandbox_url = https://{your running sandbox}/sandboxed-api`  to your gradle.properies. Make purchase and notice that sandbox API is hit

### CODE REVIEW

File: libs/iap/src/main/java/com/woocommerce/android/iap/pub/IAPSitePurchasePlanFactory.kt [hunk 1]
```diff
@@ -1,23 +1,24 @@
...
 import com.woocommerce.android.iap.internal.planpurchase.IAPPurchaseWpComPlanHandler
 import com.woocommerce.android.iap.internal.planpurchase.IAPPurchaseWpComPlanSupportCheckerImpl
+import com.woocommerce.android.iap.pub.network.IAPMobilePayAPI
 
 object IAPSitePurchasePlanFactory {
     fun createIAPSitePurchasePlan(
         context: Application,
         remoteSiteId: Long,
         logWrapper: IAPLogWrapper,
+        realMobilePayApiProvider: (String?) -> IAPMobilePayAPI,
```
Comment (kidinov) [inline libs/iap/src/main/java/com/woocommerce/android/iap/pub/IAPSitePurchasePlanFactory.kt:16]: @JorgeMucientes  FYI, there is an API change. You can look up `IAPShowcaseMobilePayAPIProvider` and maybe even copy-paste it to your implementation

Comment (JorgeMucientes) [reply libs/iap/src/main/java/com/woocommerce/android/iap/pub/IAPSitePurchasePlanFactory.kt:16]: Cool! Thanks for the heads up @kidinov

File: libs/iap/src/main/java/com/woocommerce/android/iap/pub/IAPSitePurchasePlanFactory.kt [hunk 2]
```diff
@@ -30,4 +31,18 @@ object IAPSitePurchasePlanFactory {
         val iapManager = IAPManagerFactory.createIAPManager(context, logWrapper)
         return IAPPurchaseWpComPlanSupportCheckerImpl(iapManager)
     }
+
+    private fun providerMobilePayAPI(
```
Comment (AnirudhBhat) [inline libs/iap/src/main/java/com/woocommerce/android/iap/pub/IAPSitePurchasePlanFactory.kt]: 🔍 ❓ Nitpick: I was wondering if we could write tests for this logic. This piece of code decides which server to hit and it seems like it's worth writing tests for this. WDYT?

For example, Something like this
```
class ServerProvider {

    fun provideServer(
        buildConfigWrapper: BuildConfigWrapper,
        logWrapper: IAPLogWrapper,
        realMobilePayApiProvider: (String?) -> IAPMobilePayAPI
    ) : IAPMobilePayAPI {
        return if (buildConfigWrapper.IAP_TESTING_SANDBOX_URL.isNotEmpty()) {
            realMobilePayApiProvider(buildConfigWrapper.IAP_TESTING_SANDBOX_URL)
        } else {
            if (buildConfigWrapper.debug) {
                IAPMobilePayAPIStub(logWrapper)
            } else {
                realMobilePayApiProvider(null)
            }
        }
    }
}
```

```class ServerProviderTest {

    private val logWrapperMock: IAPLogWrapper = mock()
    private val buildConfigWrapper: BuildConfigWrapper = mock()
    private var realMobileApiProvider: (String?) -> IAPMobilePayAPI = mock()

    @Before
    fun setup() {

    }

    @Test
    fun `when sandbox url is not empty, use real mobile api provider` () {
        whenever(buildConfigWrapper.IAP_TESTING_SANDBOX_URL).thenReturn("abcd")

        ServerProvider().provideServer(
            buildConfigWrapper,
            logWrapperMock,
            realMobileApiProvider
        )

        verify(realMobileApiProvider).invoke("abcd")
    }

    @Test
    fun `when sandbox url is empty and debug build, use stub mobile api provider` () {
        whenever(buildConfigWrapper.IAP_TESTING_SANDBOX_URL).thenReturn("")
        whenever(buildConfigWrapper.debug).thenReturn(true)

        ServerProvider().provideServer(
            buildConfigWrapper,
            logWrapperMock,
            realMobileApiProvider
        )

        verify(realMobileApiProvider, never()).invoke(any())
    }

    @Test
    fun `when sandbox url is empty and not debug build, use real mobile api provider` () {
        whenever(buildConfigWrapper.IAP_TESTING_SANDBOX_URL).thenReturn("")
        whenever(buildConfigWrapper.debug).thenReturn(false)

        ServerProvider().provideServer(
            buildConfigWrapper,
            logWrapperMock,
            realMobileApiProvider
        )

        verify(realMobileApiProvider).invoke(null)
    }
}
```

```
class BuildConfigWrapper {
    val debug = BuildConfig.DEBUG
    val IAP_TESTING_SANDBOX_URL = BuildConfig.IAP_TESTING_SANDBOX_URL
}
```

```
private fun providerMobilePayAPI(
        logWrapper: IAPLogWrapper,
        realMobilePayApiProvider: (String?) -> IAPMobilePayAPI
    ): IAPMobilePayAPI =
        ServerProvider().provideServer(
            BuildConfigWrapper(),
            logWrapper, realMobilePayApiProvider
        )

```

**Comment by AnirudhBhat** [comment]
Thanks for your work on this @kidinov I've reviewed the code and just left one minor nitpick that is nonblocking for the merge. 

I have asked @samiuelson 's help in testing the functionality of this PR since my IAP setup isn't working.

File: libs/iap/src/main/java/com/woocommerce/android/iap/pub/IAPSitePurchasePlanFactory.kt (hunk 2 above)
Comment (kidinov) [reply libs/iap/src/main/java/com/woocommerce/android/iap/pub/IAPSitePurchasePlanFactory.kt]: 👍

2e972fe8480d37fab11ed42afc165057c7d8816d

**Review by samiuelson — APPROVED** [review]
Works as expected!

**Pull request merged**